### PR TITLE
Implement Retrieve:codeComparator

### PR DESCRIPTION
**Pull request by dehall**
TL;DR:

This PR implements Retrieve.codeComparator, so retrieves now honor `in`, `=`, and `~` semantics for ValueSets, Codes, Concepts, and code lists, including multi-valued code properties. Exact equality is now supported and compares code displays, while equivalence continues to match code/system only. CodeSystem retrieves are intentionally unsupported but wired up so that if the appropriate terminology lookups are ever implemented, they will work here.

---

This PR implements proper support for the [codeComparator property of a Retrieve](https://cql.hl7.org/04-logicalspecification.html#codecomparator) (implements #293). After a lot of thinking, testing, and discussion about this, it turns out this PR doesn't change that much about how Retrieves are handled but it does tune the implementation to be a little more aligned to the spec, and fixes a few cases that were not on the happy path.

#### Spec / Background
For context, there are 3 key parts to a Retrieve that are relevant to this discussion. 1, the `codeProperty`, 2, the `codeComparator`, and 3, the `codes` (or `terminology`). For example, in `[Condition: bodySite in "myValueSet"]` -- `codeProperty` is `bodySite`, `codeComparator` is `in`, and `codes` is `myValueSet`. If the `codes` is specified but the `codeProperty` and `codeComparator` are not, those fields are populated in translation based on the model. (there's a default `codeProperty` for every resource type, and a default operator for whichever type `codes` is). So `codeComparator` is always present, and currently allows 3 values (`~`, `=`, and `in`).

Note from this point on I'll refer to the `codeProperty` and the value of that field on a given resource as the "Left-Hand Side" or LHS, and I'll refer to the `codes` or `terminology` that it's being compared to as the "Right-Hand Side" or RHS. So eg, `bodySite in myValueSet`, `bodySite` is the LHS and `myValueSet` is the RHS. Otherwise referring to "codes" gets too confusing. 

#### Implementation 
The core of the comparison logic is performed in `Retrieve.recordMatchesCodeOrVS` (src/elm/external.ts). Previously, this would only look at the type of the RHS to appropriately call `hasMatch` on either a single ValueSet object or on the elements of a code list. Now, this branches out into 3 separate functions for each codeComparator type, with further branches based on the type of the RHS to call the appropriate function.

Now, the happy path of this is pretty easy to see what should happen: `[Condition: code in "myValueSet"]` or `[Condition: code = "mySingleCode"]`. But it turns out that the operators in a Retrieve operate differently than they do in a standalone definition, and are much more permissive. Combinations that would be generally be illegal, eg `"mySingleCode" = "myCodeSystem"`, are allowed in a Retrieve, eg `[Condition: code = "myCodeSystem"]`. We discussed this a little bit on Zulip: [#cql > Retrieve.codeComparator implementation details](https://chat.fhir.org/#narrow/channel/179220-cql/topic/Retrieve.2EcodeComparator.20implementation.20details/with/609251783)
The takeaway from that is that it is intentional, it's meant to be flexible but also consistent for the user, but the execution engine does have to concern itself with the different combinations.

However, while the ELM translator permits every combination of operators and operands in a Retrieve, even combinations that are not allowed standalone, for valid combinations it does try to be smart about the types, and will list-promote or expand value sets in certain cases. For example:
`[Condition: code = "myValueSet"]` --> the resulting Retrieve.codes is an `ExpandValueSet` on a `ValueSetRef` and gets passed into `recordMatchesCodeOrVS` as a list of Codes
`[Condition: code in "myValueSet"]` --> the resulting Retrieve.codes is a plain `ValueSetRef` and gets passed into `recordMatchesCodeOrVS` as a ValueSet object
So my takeaway here is, we should generally consider a ValueSet and Code[] to be equivalent, if possible. 

And beyond that, it actually turns out that because of list promotion, even if in the CQL the RHS is a single Code, it will be represented in the resulting Retrieve.codes as a ToList wrapping that code. So even if the CQL says `code = myCode`, it's really `code = { myCode }`.
This means that there's no way to distinguish from the ELM whether the CQL was originally written as a single code or as a list/Concept/ValueSet that contained only a single code. So that greatly reduces the number of combinations we have to consider, but it also means that some of the trivial combinations get folded up into less obvious combinations and we have to make implementation decisions. 

Start with the `=` operator for equality. So now implementing this I see a few options:
1. Use regular list equality, LHS = RHS. `{ code1 } = { code1 }` is true, but `{ code1 } = { code1, code2 }` is false.
2. Interpret the intent of the operator as "some exact match exists between the LHS and RHS". `{ code1 } = { code1, code2 }` is now true. 
3. Different logic depending on the size of RHS. Assume that a list with a single item was list-promoted, anything larger was originally a list.

Let's consider a couple examples to help decide:

- `[Condition: code = mySingleCode]` 
  - I imagine this is probably the most common example
  - Condition.code is 0..1 so this would likely work "as expected" no matter how we implement it
- `[Condition: bodySite = mySingleCode]` - reminder Condition.bodySite is 0..*. 
  - If a given condition has 2 bodySite codes, this could become a comparison `{ code1, code2 } =  { code1 }` . Should this evaluate to true? I think so.
  - Option (1) means "exactly this list" which may be uncommon in a retrieve. (1) would return false in this example.
- `[Condition: bodySite = myConcept]`
  - Again bodySite is 0..*, and a concept may have multiple codes. List equality as (1) means "exactly this list" which could be a valid use case, but I expect most of the time that's not what's expected.
- `[Condition: bodySite = myCodeList]`
  - What if someone wanted to search for a Condition with exactly a specific list of codes, ie, not just a single code but multiple? Regular list equality as in (1) would allow this possibility, but (2) does not.

Approach (3) might produce a better result in some of these cases, but would also result in inconsistent and potentially unexpected behavior in others, so my preference is not to go that route. 
So that leaves (2). More permissive than (1), there's no example that should return false in (2) but true in (1).

#### Final Implementation Approach
So after all that, I implemented the operators with the interpretation that `~` and `=` mean "some element from the LHS matches some element from the RHS", and not "the entirety of the LHS matches the entirety of the RHS". [`in` for CodeSystems](https://cql.hl7.org/09-b-cqlreference.html#in-codesystem) and [`in` for ValueSets](https://cql.hl7.org/09-b-cqlreference.html#in-valueset) are both defined to have equivalence semantics, so `~` and `in` are essentially the same, and both implement essentially the same logic as previously existed. So the only thing that is really "new" here is support for the `=` operator, as well as more specific Error messages when calling the few illegal combinations.

#### Supporting Changes

- In order for any comparison using `=` to work, the code display needs to be populated, but previously there were some paths that copy code objects without copying the display. Those were fixed here, and various test fixtures and test cases were updated as well. (Possibly more than needed but only for consistency within a single file, eg, the valuesets fixture)
- The logic that looks up the `codeProperty` had been assuming the field was a single code, but this is not necessarily true and there are many fields that are a list of codes, eg `Condition.bodySite` which is a 0..* CodeableConcept. Ideally there would be a data-model-aware generic getter, but for now that doesn't exist. I wasn't sure whether it was better to update `getCode` to possibly return a list, so I added a new function `getCodeOrCodes` which always returns a list even if getting a single code.
- Added a `CodeSystem.hasMatch` function that just throws an error for now. The InCodeSystem operation is not supported yet (https://github.com/cqframework/cql-execution/issues/289) and too large to implement here, and it felt cleaner to add that "not yet implemented" error in a `CodeSystem` member function rather than the function handling the `in` operator. However, it turns out the code execution flow doesn't make it this far, since earlier on the code tries to evaluate the CodeSystem as if it were a ValueSet and that lookup fails. I didn't modify that earlier step since there's no real way to make it work without a full implementation, so the impact is just that the error message might not be the most ideal for the situation.

Pull requests into cql-execution require the following.
Submitter and reviewer should ✔ when done.
For items that are not-applicable, mark "N/A" and ✔.

**Submitter:**

- [x] This pull request describes why these changes were made
- [x] Code diff has been done and been reviewed (it does not contain: additional white space, not applicable code changes, debug statements, etc.)
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass
- [x] Code coverage has not gone down and all code touched or added is covered.
- [x] Code passes lint and prettier (hint: use `npm run test:plus` to run tests, lint, and prettier)
- [ ] All dependent libraries are appropriately updated or have a corresponding PR related to this change
- [x] `cql4browsers.js` built with `npm run build:browserify` if source changed.

**Reviewer:**

Name:

- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

### CODE REVIEW

File: src/elm/external.ts [hunk 1]
```diff
@@ -79,10 +83,50 @@ export class Retrieve extends Expression {
   }
 
   recordMatchesCodesOrVS(record: any, codes: any) {
-    if (typeIsArray(codes)) {
-      return (codes as any[]).some(c => c.hasMatch(record.getCode(this.codeProperty)));
+    const recordCodeValue = record.getCodeOrCodes(this.codeProperty);
```
Comment (cmoesel) [inline src/elm/external.ts]: Any existing implementations will only support `getCode`. To keep this backwards-compatible, I suggest we check `typeof record.getCodeOrCodes === 'function'` and use it if it passes, but fall back to `record.getCode` otherwise.

File: src/elm/external.ts [hunk 2]
```diff
@@ -79,10 +83,50 @@ export class Retrieve extends Expression {
...
+    switch (this.codeComparator) {
+      case 'in':
+        return this._in(recordCodeValue, codes);
+
+      case '~':
+        return this._equivalent(recordCodeValue, codes);
+
+      case '=':
+        return this._equal(recordCodeValue, codes);
+    }
```
Comment (cmoesel) [inline src/elm/external.ts:112]: I know that at one point, it was possible for ELM retrieves _not_ to have a `codeComparator` -- and that was just fine with us since we never looked at it anyway. In addition, the logical model for Retrieve lists `codeComparator` as `0..1`, so a missing `codeComparator` is still valid according to spec. In that case, I think we should add a sensible `default:` to the `switch` statement. I was going to suggest to default to `in` or `~` respectively depending on if `codes` is an array or not; but based on your PR description, it sounds like it's always an array, so maybe just default to `in`?

Or maybe we need to keep a path for the code property not to return an array -- because the fact that the CQL-to-ELM translator auto-converts it to a list kind of sounds like an implementation detail to me. Is there anywhere in the spec that says it _has_ to do that? Might other implementations (or a future version) _not_ do that?

File: src/elm/external.ts [hunk 3]
```diff
@@ -79,10 +83,50 @@ export class Retrieve extends Expression {
...
+  }
+
+  _in(lhs: Code[], rhs: any) {
```
Comment (cmoesel) [inline src/elm/external.ts]: I don't think we generally use `_` prefixes to distinguish "private" methods in most of our classes. We do it in `patient.js` since that served as a model for the PatientSource API and we wanted it _really_ clear what was part of the API and what wasn't, but we don't generally use that convention elsewhere. I'd suggest removing the `_` prefix on this and the other methods. Since we're in TypeScript, you can use the `private` modifier instead if you'd like. We haven't been very consistent about using that, but I think it's a good idea.

File: src/elm/external.ts [hunk 4]
```diff
@@ -79,10 +83,50 @@ export class Retrieve extends Expression {
...
+      // for code lists, fallback to the implementation in ~
+      return this._equivalent(lhs, rhs);
     }
   }
+
+  _equivalent(lhs: Code[], rhs: any) {
+    if (rhs instanceof CodeSystem) {
+      throw new Error("Operator '~' is not defined for Code ~ CodeSystem");
+    } else if (rhs instanceof ValueSet) {
+      // TODO: explain
```
Comment (cmoesel) [inline src/elm/external.ts]: I'm guessing that you intended to add the explanation as part of this PR (not for a future task), right?

File: src/types/cql-patient.interfaces.ts [hunk 5]
```diff
@@ -17,6 +17,7 @@ export interface RecordObject {
   get(field: any): any;
   getId(): any;
   getCode(field: any): any;
+  getCodeOrCodes(field: any): Code[] | undefined;
```
Comment (cmoesel) [inline src/types/cql-patient.interfaces.ts]: We should probably start with this as optional for backwards-compatibility. Otherwise we need to do a major version bump and break existing implementations.

File: test/elm/external/data.cql [hunk 6]
```diff
@@ -1,20 +1,65 @@
...
 
-codesystem "SNOMED": '2.16.840.1.113883.6.96'
+codesystem "SNOMED": '2.16.840.1.113883.6.96' version '2013-09'
 valueset "Ambulatory/ED Visit": '2.16.840.1.113883.3.464.1003.101.12.1061'
 valueset "Annual Wellness Visit": '2.16.840.1.113883.3.526.3.1240'
 code "Viral pharyngitis code": '1532007' from "SNOMED" display 'Viral pharyngitis (disorder)'
 concept "Viral pharyngitis": { "Viral pharyngitis code" } display 'Viral pharyngitis (disorder)'
+
+valueset "Body locations": '2.16.840.1.113762.1.4.1181.77' // Pain body location reference set
+code "Nose code": '45206002' from "SNOMED" display 'Nasal structure (body structure)'
```
Comment (cmoesel) [inline test/elm/external/data.cql:11]: I love me a good nose code.

File: test/elm/external/data.cql [hunk 7]
```diff
@@ -1,20 +1,65 @@
...
+
+// duplicated definitions purely for readability of the large cross-product of test cases
+codesystem "myCodesystem": '2.16.840.1.113883.6.96' version '2013-09'
+valueset "myValueSet": '2.16.840.1.113883.3.464.1003.102.12.1011'
+code "myCode": '1532007' from "myCodesystem" display 'Viral pharyngitis (disorder)'
+concept "myConcept": { "myCode" } display 'Viral pharyngitis (disorder)'
+define myCodeList: { myCode }
```
Comment (cmoesel) [inline test/elm/external/data.cql]: This is minor (or perhaps even trivial), but it is considered best practice in CQL to name code systems and value sets after the "friendly name" of the code system and title of the value set. See [Using CQL w/ FHIR IG](https://build.fhir.org/ig/HL7/cql-ig/en/using-cql.html#code-systems).

I can see here why it's easier to understand the other expressions (and avoid mistakes) if you put the type of thing (code system, value set, code, concept) in the name though. I think that's OK, but generally speaking, most CQL identifiers also start with a capital letter, so consider `MyCodeSystem`, `MyValueSet`, etc.

**Review by cmoesel — CHANGES_REQUESTED** [review]
Thanks for the helpful description and context for this PR. You've thought through this well and I think the approach makes sense. I've made a few suggestions for your consideration. My biggest concern is that I don't think we need to break backward compatibility for this, so I'd like to adjust where necessary to ensure existing solutions (and ELMs) continue to work.

File: src/elm/external.ts (hunk 4 above)
Comment (dehall) [reply src/elm/external.ts]: Ack, yes sorry I should have caught that before the PR

File: src/elm/external.ts (hunk 1 above)
Comment (dehall) [reply src/elm/external.ts]: I'd even be fine removing `getCodeOrCodes` altogether, but we should probably talk about what the best approach is for when it's not available. This is "supporting change" 2 I mentioned at the bottom of the PR description, where we need some way of getting the codes from a list. `record.getCode` as we have it implemented here will only get a single code, and only if the referenced field is a single code. i.e, it won't extract the first code from a list or anything like that. So I think my fallback has to be plain `get`

Comment (cmoesel) [reply src/elm/external.ts]: I'm not opposed to introducing `getCodeOrCodes` as long as it is optional. But we could also update the signature of `getCode` to allow for returning a single code _or_ array of codes. I know `getCode` isn't plural, but neither is `codeProperty`, despite the fact it can potentially return multiple codes. You can also use `get` -- but if you look at the simple implementation we have in `cql-execution`, you'll notice that [get](https://github.com/cqframework/cql-execution/blob/master/src/cql-patient.ts#L42-L57) calls `getCode`. That said, if you look at cql-exec-fhir (which is the one we care about), you'll see that its [getCode](https://github.com/cqframework/cql-exec-fhir/blob/master/src/fhir.js#L217-L223) calls `get`. So... 🤷. 

On second thought, maybe `get` doesn't work, because we need a CQL System Code, and the data model may return its own representation of a code (e.g., cql-exec-fhir will return a `FHIR.code`, `FHIR.coding`, or `FHIR.CodeableConcept`) -- and translating that to a CQL Code is a data-model-specific implementation that doesn't belong in `cql-execution`.

Comment (dehall) [reply src/elm/external.ts]: Tracing through the cql-exec-fhir version, looks like that will already return either an array or a single code, so ok, I'll go forward with that approach. Theoretically any other data sources that don't support reading from an array won't work correctly for that example `[Condition: bodySite = Something]`, but I guess they already wouldn't work today for that anyway.

File: test/elm/external/data.cql (hunk 7 above)
Comment (dehall) [reply test/elm/external/data.cql]: Updated and made the comment at the start of this section bigger and bolder to say "these names are intentionally not best practice so don't use this as an example" https://github.com/cqframework/cql-execution/pull/370/commits/25e0862cafada5db6ffd5ac39dafbaf4497d3941

File: src/types/cql-patient.interfaces.ts (hunk 5 above)
Comment (dehall) [reply src/types/cql-patient.interfaces.ts]: Removed entirely in https://github.com/cqframework/cql-execution/pull/370/commits/1d2a4fec56b86886cc2ed652d246bf3c6dff4403

File: src/elm/external.ts (hunk 4 above)
Comment (dehall) [reply src/elm/external.ts]: Added some explanation for this here and in a couple other places - essentially a shorter version of this PR's long description. If someone needs more they can git blame and find this PR. https://github.com/cqframework/cql-execution/pull/370/commits/0ec159b8e7a3c46da6128bbb9ff2f18b8b0e24d7

File: src/elm/external.ts (hunk 3 above)
Comment (dehall) [reply src/elm/external.ts]: Updated to use `private` keyword in https://github.com/cqframework/cql-execution/pull/370/commits/b80d37e8f7e6016a86e96d016d03f1ebabba9466

File: src/elm/external.ts (hunk 1 above)
Comment (dehall) [reply src/elm/external.ts]: Updated so this now calls `getCode`, which will return either a single code or a list of codes. https://github.com/cqframework/cql-execution/pull/370/commits/c5d19949d2791f26b0638a5e6ee3b424e1e3084c

File: src/elm/external.ts (hunk 2 above)
Comment (dehall) [reply src/elm/external.ts:112]: `codeComparator` can definitely be unset, eg `[Condition]`, but if there is a terminology filter on the retrieve then all 3 of `codes`, `codeProperty`, and `codeComparator` should always be set. Note the code was previously and still assumes that `codeProperty` is set, but it's also 0..1.

Re: `codes` always being a list -- I'm pretty sure it will always be a list based on the spec here:
> The codes element optionally specifies an expression that results in a List\<Code\> to match against. Only those clinical statements that match at least one of the specified codes will be returned.  

(https://cql.hl7.org/04-logicalspecification.html#codes)
and also, just from the name being "codes". If it could be something else I think that would have to be made more explicit.

All that said, it's easy to add some extra guardrails just in case something unexpected happens. Added a default to the switch that calls "in" and extra logic in the 3 handler functions that handles a single code being passed in instead of list. Only thing is, for test code coverage, I don't expect these lines to be possible to hit by executing translated CQL from our tests. I can wire up some tests by calling the functions directly if you think that would useful

Comment (cmoesel) [reply src/elm/external.ts:112]: I think maybe I was confusing the codes from the retrieve with the codes from the data model. I agree that it sounds like we can count on the codes from the retrieve always being a list (or at least resolving to a list, as is the case for value sets). It's the codes from the data model that could be either one code or a list -- and _that's_ what we need to be sure to handle. I apologize for the confusion.

Comment (cmoesel) [reply src/elm/external.ts:112]: Or maybe I wasn't confused. Maybe it's just the spec that's confusing -- because if it's always a list, then what really is the difference between `~` and `in`? They both mean "use equivalence to see if one of the codes from the data matches one of the codes from the retrieve."

Comment (dehall) [reply src/elm/external.ts:112]: Yes I think it's the spec that's confusing. 

The intent was that you could use these 3 operators with the appropriate operands and they would work the same way inside a retrieve as outside. (Per Bryn [#cql > Retrieve.codeComparator implementation details](https://chat.fhir.org/#narrow/channel/179220-cql/topic/.E2.9C.94.20Retrieve.2EcodeComparator.20implementation.20details/near/609223706) ) And in practice I'm sure authors will always use `in` if it's a VS, CS, or list, or `~`/`=` if it's a single code. But because of list promotion and the fact that every combination of operator and operand types is legal, we have to figure out how to handle these other possibilities that could technically appear.

File: src/elm/external.ts [hunk 8]
```diff
@@ -79,10 +83,87 @@ export class Retrieve extends Expression {
...
+    if (rhs instanceof CodeSystem) {
+      throw new Error("Operator '~' is not defined for Code ~ CodeSystem");
+    } else if (rhs instanceof ValueSet) {
+      // It's not obvious that this combination should even be allowed.
+      // If in the CQL representation of the Retrieve, the terminology is a ValueSet,
+      // eg [Condition: code ~ "SomeValueSet"],
+      // the CQL-to-ELM translation may turn that into either:
+      //  - a `ValueSetRef` expression (appears here as a `ValueSet`)
+      //  -  or an `ExpandValueSet` expression wrapping the `ValueSetRef` (appears here as a `Code[]`).
+      // Because we can't tell the original intent here, we treat the ValueSet as if it were a Code[].
```
Comment (cmoesel) [inline src/elm/external.ts]: I don't think I understand what is being conveyed here. It says a `ValueSetRef` appears as a `ValueSet` and `ExpandValueSet(ValueSetRef)` appears as a `Code[]`. It then says it can't tell intent and treats `ValueSet` as a `Code[]`, but... `rhs` is already either a `ValueSet` or a `Code[]` here, so... I don't understand the last sentence that makes is sound like there is an ambiguity we are choosing to resolve here.

And when you say `appears here as`, are you referring to the type of `rhs`? I'm assuming so, but I'm not sure.

If it's easier, I'd be OK with an even shorter note with a link to the Zulip conversation and/or this PR.

**Review by cmoesel — CHANGES_REQUESTED** [review]
This is looking good and most of it makes sense to me. The only thing I struggled with was the actual comment about the retrieve implementation. See my individual comment for more.

File: src/elm/external.ts (hunk 8 above)
Comment (dehall) [reply src/elm/external.ts]: I'll try to rework this tomorrow. I think this needs to be combined with the comment in the function below it. 

The point of the comment to explain why `code ~ ValueSet` is allowed, because to me that's surprising. My initial reaction was that combination should not be allowed. (And `code ~ CodeSystem` is disallowed just above it so why not `ValueSet`?)

There are two ambiguities and taken together it means, when RHS is a `Code[]` here, we don't know whether it was originally written in the CQL as a ValueSet reference that got expanded into a `Code[]`, or if it was already written as a Code List, or if it was written as a single `Code` that got list-promoted into a `Code[]`. Because we don't know the intent, we kind of always have to allow the `~` operator when RHS is a `Code[]`. 
Then to avoid inconsistent behavior, we should allow `~` when RHS is a `ValueSet` object as well because otherwise we'd be allowing it in the situations that pre-expand it into a `Code[]` but not in the situations that don't. It's definitely awkward

Comment (dehall) [reply src/elm/external.ts]: How about this:

> It's not obvious that `code ~ ValueSet` and `code ~ Code[]` should be allowed, when `code ~ CodeSystem` is disallowed. Same for operator `=` below.
> In short, due to various translator behaviors, when `rhs` here is `Code[]`, it's possible it was originally written in the CQL as a Code, a List\<Code\>, or a ValueSet. We can't distinguish, so we have to allow all of them.
> Then, again because of translator behaviors, `x ~ ValueSet` could reach this point with `rhs` either being a `Code[]` or a `ValueSet` object. Because we can't filter it out in the `Code[]` path, we should allow it in the `ValueSet` path as well to reduce unexpected inconsistencies.
> See full description and discussion at (link to this PR)

Comment (cmoesel) [reply src/elm/external.ts]: Yes, that description clarifies it very well. Let's go with that. Thanks!

Comment (dehall) [reply src/elm/external.ts]: Updated in https://github.com/cqframework/cql-execution/pull/370/commits/9468349a317e72e428145da1d84933084fc917cd

**Review by cmoesel — APPROVED** [review]
This looks good to me. I regressed it against a bunch of old CDS Connect artifacts and all the tests pass. Of course I'd like confirmation that it passes fqm-execution regression too before merging.

**Review by hossenlopp — APPROVED** [review]
Tested in fqm-execution against a large set of CMS Measures. No changes to calculation as expected.
